Add time_series_metric parameter

csoulios opened the pull request:
This PR adds the `time_series_metric` parameter to the following field types:

- Numeric field types
- `histogram`
- `aggregate_metric_double`

The `time_series_metric` parameter is of type string (default is null) and is used to mark that a field is a time series metric field with a specific metric type. The following metric types are supported for each Elasticsearch field type: 

|Metric type|ES field type
|---|---
|gauge|number, `aggregate_metric_double`
|counter|number, `aggregate_metric_double`
|histogram|`histogram`
|summary|`aggregate_metric_double`


Example of setting the `time_series_metric` parameter to a field can be seen here:
```
PUT my-index-000001
{
  "mappings": {
    "properties": {
      "free_memory": {
        "type": "double",
        "time_series_metric": "gauge"
      },
      "net_bytes_in": {
        "type": "double",
        "time_series_metric": "counter"
      }
    }
  }
}
```

This PR together with https://github.com/elastic/elasticsearch/pull/74450 annotate which Elasticsearch fields are time series dimensions or metrics.

Closes https://github.com/elastic/elasticsearch/issues/74014

csoulios committed:
Added metric mapping parameter for Number fields

csoulios committed:
Moved metric param definition to a static method

So that it's reusable.
Added more tests

csoulios committed:
Add metric param to aggregate_metric_double fields

csoulios committed:
Add metric param to histogram fields

csoulios committed:
Renamed metric param to time_series_metric

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

csoulios committed:
Checkstyle

imotov reviewed (commented):
It looks good to me in general. I have only a couple of comments. I think some integration level tests, maybe even rest test could be helpful, so we have some clear examples how this is used. It also looks like long fields can be metrics and dimensions at the same time, not sure if this is intentional. It would be great if @nik9000 could review it since he is working with this at the moment.

nik9000 commented:
> It would be great if @nik9000 could review it since he is working with this at the moment.

I'm not 100% sure we'll keep allowing `long` to be a dimension. But I guess it's easy enough to add the assertion that you can't do that so we may as well.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -126,6 +138,25 @@ public Builder(String name, NumberType type, ScriptCompiler compiler, boolean ig
+            final Consumer<String> defaultValidator = metric.getValidator();
+            metric.setValidator(v -> {
+                // Call the default validator first
+                defaultValidator.accept(v);
```
@romseygeek do you think this should be easier? Or the default? Or something?

romseygeek replied:
Maybe add a `wrapValidator` method that does this for you?

csoulios replied:
I was thinking that perhaps we could add a list of validators for a parameter. Those validators can be executed sequentially. In this case a parameter can have its validator and its subclasses can add more.

It's an easy change and I can submit a different PR for that. WDYT @romseygeek?

romseygeek replied:
Sounds good to me.

nik9000 commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/range/RangeAggregatorTests.java:
```diff
@@ -146,7 +146,8 @@ public void testUnboundedRanges() throws IOException {
                 null,
-                false
+                false,
+                null
```
So so so many parameters. Nothing you can do about it now, but we should think about something cleaner eventually.

csoulios replied:
I agree it's ugly. I can create a new constructor with default values (`false` and `null`) for the time series parameters. WDYT?

nik9000 replied:
I'm not at all sure! Maybe a builder or some test utility methods or something.

I don't think it's a problem for this PR though. It seems like a great candidate for a follow up change.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -83,8 +85,18 @@ private static NumberFieldMapper toType(FieldMapper in) {
+         * For the numeric fields gauge and counter metric types are
+         * supported
+         */
+        private final Parameter<String> metric;
```
Should this be `Parameter<MetricType>`? `Setting`s can do things like this. I'm not sure about `Parameter`.

csoulios committed:
Merge branch 'master' into metric-mapping-param

csoulios committed:
Merge branch 'master' into metric-mapping-param

csoulios committed:
Add new validator to t_s_metric for numeric fields

Now that multiple validators are supported (#77073) we can add
a new validator instead of replacing the default param validator.

csoulios committed:
Changed metric params from String to enum

csoulios committed:
Merge branch 'master' into metric-mapping-param

csoulios commented on server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -906,6 +907,68 @@ protected void toXContent(XContentBuilder builder, boolean includeDefaults) thro
+         * @param enumClass     the enumeration class the parameter takes values from
+         * @param values        the set of values that the parameter can take
+         */
+        public static <T extends Enum<T>> Parameter<T> restrictedEnumParam(
```
@romseygeek can you please have a look at this and let me know what you think?

I thought this could be re-usable

csoulios committed:
Merge branch 'master' into metric-mapping-param

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -905,6 +906,68 @@ protected void toXContent(XContentBuilder builder, boolean includeDefaults) thro
+            T defaultValue,
+            Class<T> enumClass
+        ) {
+            Set<T> acceptedValues = EnumSet.noneOf(enumClass);
```
`allOf`?

csoulios replied:
Of course `allOf`. That was a silly left over from a copy-paste. Thanks

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -905,6 +906,68 @@ protected void toXContent(XContentBuilder builder, boolean includeDefaults) thro
+                    @SuppressWarnings("unchecked")
+                    T enumValue = Enum.valueOf(enumClass, (String) o);
+                    return enumValue;
+                } catch (Exception e) {
```
Maybe just `IllegalArgumentException`? I think I'm just being paranoid, but still.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesParams.java:
```diff
@@ -0,0 +1,45 @@
+        gauge,
+        counter,
+        histogram,
+        summary
```
I'd make them `GUAGE` and `COUNTER` and `HISTOGRAM` just to follow along with normal java stuff. You'd have to convert to upper case on parse but I think that's fine.

csoulios replied:
I started writing this change, but it turned out that code became a lot more complicated just to convert input to upper case and also convert enum values to lower case when logging error messages. Otherwise, user receives error messages like the following:

`"Unknown value [histogram] for field [time_series_metric] - accepted values are [GAUGE, COUNTER]"` 

So, sticking to enum constants same as the expected input values would result in a a lot simpler code.

We have followed the same approach [here](https://github.com/elastic/elasticsearch/blob/5baabff6670a8ed49297488ca8cac8ec12a2078d/x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java#L93-L101)

nik9000 replied:
Oh well. I don't like it there either! But I think it's cool - they can stay lowercase.

csoulios replied:
It can also become relatively simple  if I replace the generic implementation (`restrictedEnumParam()`) from `FieldMapper` class and only add the custom implementation for the `MetricType` enum in the `TimeSeriesParams`.

Is a more specific to `MetricType` implementation with capitalized names  a better approach here?

nik9000 replied:
Sorry I don't understand - do you mean that you can either share the enum infrastructure or capitalize the names of the constants? If that's the case I think I prefer sharing the enum infrastructure. That way if we come up with a clever way to capitalize the names we can get it in a follow up and hit everything using the infrastructure.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -905,6 +906,68 @@ protected void toXContent(XContentBuilder builder, boolean includeDefaults) thro
+            T defaultValue,
+            Class<T> enumClass,
+            Set<T> values
+        ) {
```
It's probably worth adding a test to `ParameterizedMapperTests`.

nik9000 reviewed (commented):
I'm pretty happy with it. I left some small questions. @romseygeek, what do you think about that shared stuff for enums?

romseygeek commented:
> @romseygeek, what do you think about that shared stuff for enums?

I'm ok with it, I think. Does it gain us a lot over just having a String param and then using `Enum.valueOf()` to resolve it though? Maybe a nicer error message.

nik9000 commented:
> I'm ok with it, I think. Does it gain us a lot over just having a String param and then using `Enum.valueOf()` to resolve it though? Maybe a nicer error message.

I liked it because it gave parity with settings and stuff.

csoulios committed:
Addressed reviewer comments

csoulios committed:
Added tests

nik9000 commented on server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:
```diff
@@ -260,6 +261,59 @@ public void testDimension() throws IOException {
         assertThat(e.getCause().getMessage(), containsString("Parameter [dimension] cannot be set"));
     }
 
+    protected <T> void assertMetricType(String metricType, Function<T, Enum<TimeSeriesParams.MetricType>> checker)
```
Do you think it'd be worth sticking this in the superclass so we don't have to copy and paste it?

csoulios committed:
Moved assertMetricType() to parent class

csoulios committed:
Merge branch 'master' into metric-mapping-param

csoulios committed:
Merge branch 'master' into metric-mapping-param

nik9000 reviewed (approved):
LGTM.

Sorry for the delay!
